lms/djangoapps/courseware/module_render.py

```diff
@@ -486,6 +506,51 @@ def publish(block, event_type, event):
             with tracker.get_tracker().context(event_type, context):
                 track_function(event_type, event)
 
+    def handle_completion_event(block, event):
+        """
+        Submit a completion object for the block.
+        """
+        BlockCompletion.objects.submit_completion(
+            user=user,
+            course_key=course_id,
+            block_key=block.scope_ids.usage_id,
+            completion=event['completion'],
+        )
+
+    def handle_grade_event(block, event):
+        """
+        Submit a grade for the block.
+        """
+        SCORE_PUBLISHED.send(
+            sender=None,
+            block=block,
+            user=user,
+            raw_earned=event['value'],
+            raw_possible=event['max_value'],
+            only_if_higher=event.get('only_if_higher'),
+            score_deleted=event.get('score_deleted'),
+        )
+
+    def handle_deprecated_progress_event(block, event):
+        """
+        DEPRECATED: Submit a completion for the block represented by the
+        progress event.
+
+        This exists to support the legacy progress extension used by
+        edx-solutions.  New XBlocks should not emit these events, but instead
+        emit completion events directly.
+        """
+        requested_user_id = event.get('user_id', user.id)
+        if requested_user_id != user.id:
+            log.warning("{} tried to submit a completion on behalf of {}".format(user, requested_user_id))
+            return
+        BlockCompletion.objects.submit_completion(
```

Comment:
@jcdyer @bradenmacdonald: IMPORTANT: Before this or any open PRs merge, can we please add the Waffle flag to ensure that no completion events will be written as these PRs land.  That means protecting this method, and possibly `handle_completion_event`, depending on where you add the protection.

Reply:
@robrap Good thing you said something now as this PR is about to merge. I don't remember that part of the discussion. Why don't we want to start collecting data now? Is it because of the unknown performance implication?

Reply:
There are a couple of reasons:
1. There are lots of dependencies between these various PRs, and we will definitely be writing partial data, and could even be writing bad data, if we don't turn it on after the work is complete.
2. Migrations, if needed, could be more of an issue with data.
3. I don't think there will be performance issues, but we don't know yet.

So, I think it is great that we are doing the development piecemeal, and agreeing to move certain cases to later stories, but a flag will allow us to know everything is properly complete (no pun intended) before we actually turn this on. :)

Reply:
Ok, that makes sense!

Reply:
@jcdyer: When we do add the flag, it may make sense to sprinkle some tests with the flag turned off to ensure that no data is written.